cerebrium/container-images/defining-container-images.mdx

```diff
@@ -0,0 +1,355 @@
+[cerebrium.deployment]
+python_version = 3.11
+```
+
+The Python version affects the entire dependency chain. For instance, some packages may not support newer Python versions immediately after release.
+
+<Warning>
+  Changes to the Python version trigger a full rebuild since they affect both
+  the base environment and all Python package installations.
+</Warning>
+
+### Adding Python Packages
+
+Python dependencies can be managed directly in TOML or through requirement files. The system caches packages at the node level to speed up builds:
+
+```toml
+[cerebrium.dependencies.pip]
+torch = "==2.0.0"
+transformers = "==4.30.0"
+numpy = "latest"
+```
+
+Or using an existing requirements file:
+
+```toml
+[cerebrium.dependencies.paths]
+pip = "requirements.txt"
+```
+
+<Tip>
+  For GitHub repositories, use shell commands instead of pip dependencies to
+  ensure proper versioning.
+</Tip>
+
+The system implements an intelligent caching strategy at the node level. When an app is built, all pip packages are cached with their exact versions, including wheel files and compiled binaries. This means subsequent builds only need to install new or updated packages, significantly reducing build times.
+
+GitHub repositories require special handling - they should be installed via shell commands rather than pip dependencies to ensure proper version control and build reproducibility.
+
+### Adding APT Packages
+
+System-level packages provide the foundation for many ML apps, handling everything from image processing libraries to audio codecs. These can be added to the `cerebrium.toml` file under the `[cerebrium.dependencies.apt]` section as follows:
+
+```toml
+[cerebrium.dependencies.apt]
+ffmpeg = "latest"
+libopenblas-base = "latest"
+libomp-dev = "latest"
+```
+
+For teams with standardized system dependencies, text files can be used instead by adding the following to the `[cerebrium.dependencies.paths]` section:
+
+```toml
+[cerebrium.dependencies.paths]
+apt = "deps_folder/pkglist.txt"
+```
+
+Since APT packages modify the system environment, any changes to these dependencies trigger a full rebuild of the container image. This ensures system-level changes are properly integrated but means builds will take longer than when modifying Python packages alone.
+
+### Conda Packages
+
+Conda excels at managing complex system-level Python dependencies, particularly for GPU support and scientific computing:
+
+```toml
+[cerebrium.dependencies.conda]
+cuda = ">=11.7"
+cudatoolkit = "11.7"
+opencv = "latest"
+```
+
+Teams using conda environments can specify their environment file:
+
+```toml
+[cerebrium.dependencies.paths]
+conda = "conda_pkglist.txt"
+```
+
+Like APT packages, Conda packages often modify system-level components. Changes to Conda dependencies will trigger a full rebuild to ensure all binary dependencies and system libraries are correctly configured. Consider batching Conda dependency updates together to minimize rebuild time.
+
+## Setup tasks
```

Comment:
I would not call it this but rather shell commands

Reply:
This is confusing, because one is called `shell_commands` and `pre_build_commands`. The title you've given makes sense but they function differently and one of them is called "Shell Commands"